Add Amazon Elastic File System (efs) module

ryansydnor opened the pull request:
##### ISSUE TYPE
 - New Modules Pull Request

##### COMPONENT NAME
efs (elastic file system)
efs_facts (elastic file system facts)

##### ANSIBLE VERSION
Tested against: ansible 2.1.1.0

##### SUMMARY
Adds the ability to create, manage, and delete elastic file systems

gregdek commented:
Thanks @ryansydnor for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

ryansydnor commented:
@gregdek - are the travis errors show stoppers?

gregdek commented:
Looks like we're trying to do a simple boto3 detection in travis and failing, so I would say no.

sivel commented:
Is the `boto` import needed?  New modules are not permitted to have a dependency on boto2.

ryansydnor commented:
No. I was testing on an older version (1.9.5) and mixing boto/boto3 to connect. I've removed all dependencies on boto and tested against ansible 2.1 and boto3. Looks good!

ryansydnor commented:
@gregdek - I believe we can remove the "needs_revision" label and add "community_review" (or equivalent).

gregdek commented:
Thanks @ryansydnor :)

jonhadfield commented:
Hi @ryansydnor 

EFS has just gone live in eu-west-1 so it would be good to get this functionality added. 

I'm testing this against latest devel:(https://github.com/ansible/ansible/commit/216a845244a29c09066fcd4032eb1bb6e487b97a)
and on the first run I get the error below. The next run succeeds and subsequent runs are successfully idempotent. Deletion also works successfully. Will try and debug if I get time.

```
Traceback (most recent call last):
  File "/var/folders/00/drbkx48d6cjd_pmvdnd3dq1w0000gn/T/ansible_7BJtvE/ansible_module_efs.py", line 689, in <module>
    main()
  File "/var/folders/00/drbkx48d6cjd_pmvdnd3dq1w0000gn/T/ansible_7BJtvE/ansible_module_efs.py", line 659, in main
    changed = connection.converge_file_system(name=name, tags=tags, targets=targets) or changed
  File "/var/folders/00/drbkx48d6cjd_pmvdnd3dq1w0000gn/T/ansible_7BJtvE/ansible_module_efs.py", line 432, in converge_file_system
    **targets[sid]
  File "/usr/local/var/pyenv/versions/2.7.10/lib/python2.7/site-packages/botocore/client.py", line 278, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/var/pyenv/versions/2.7.10/lib/python2.7/site-packages/botocore/client.py", line 572, in _make_api_call
    raise ClientError(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (IncorrectFileSystemLifeCycleState) when calling the CreateMountTarget operation: None

fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "efs"}, "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/00/drbkx48d6cjd_pmvdnd3dq1w0000gn/T/ansible_7BJtvE/ansible_module_efs.py\", line 689, in <module>\n    main()\n  File \"/var/folders/00/drbkx48d6cjd_pmvdnd3dq1w0000gn/T/ansible_7BJtvE/ansible_module_efs.py\", line 659, in main\n    changed = connection.converge_file_system(name=name, tags=tags, targets=targets) or changed\n  File \"/var/folders/00/drbkx48d6cjd_pmvdnd3dq1w0000gn/T/ansible_7BJtvE/ansible_module_efs.py\", line 432, in converge_file_system\n    **targets[sid]\n  File \"/usr/local/var/pyenv/versions/2.7.10/lib/python2.7/site-packages/botocore/client.py\", line 278, in _api_call\n    return self._make_api_call(operation_name, kwargs)\n  File \"/usr/local/var/pyenv/versions/2.7.10/lib/python2.7/site-packages/botocore/client.py\", line 572, in _make_api_call\n    raise ClientError(parsed_response, operation_name)\nbotocore.exceptions.ClientError: An error occurred (IncorrectFileSystemLifeCycleState) when calling the CreateMountTarget operation: None\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
```

I'm using the following playbook:

```
---
- name: Test EFS Module - Create
  hosts: localhost
  connection: local
  gather_facts: False
  tasks:
  - name: create efs
    efs:
      state: present
      name: my_efs
      #tags:
      #    Name: efs_name
      targets:
         - SubnetId: subnet-97bc71e1
           SecurityGroups: ["sg-60197c07"]
         - SubnetId: subnet-55eb3131
           SecurityGroups: ["sg-60197c07", "sg-cd197caa"]
    register: my_efs_details

  - name: efs creation output
    debug: var=my_efs_details

- name: Test EFS Module - Delete
  hosts: localhost
  connection: local
  gather_facts: False
  tasks:
  - name: delete efs
    efs:
      state: absent
      name: my_efs
    register: my_efs_deletion_details

  - name: efs deletion output
    debug: var=my_efs_deletion_details
```

ryansydnor commented:
Hi @jonhadfield!

First of all, thanks so much for testing! 

Unfortunately, I'm not going to be able to investigate as I'm on vacation with limited access to computers until August. 

I wonder if the API changed between beta and release leading to that exception, as it did not occur for me while testing (I have a similar set of test playbooks). I'd also be curious if we tested with different versions of boto3 (this PR could potentially be a culprit : https://github.com/boto/botocore/commit/36df38dba777ec3ab1c8c468b4b273edee09e64)

I'd greatly appreciate if you could help me look in to it. Completely understand if you cannot, but expect no progress on this PR from me for about a month.

jonhadfield commented:
The issue is that you're calling 'create_mount_target()' without checking the state being available first.
You only exit 'create_file_system()' with state of the filesystem being equal to available if wait is set to True. You'll see I didn't have wait set in my example playbooks, so it defaulted to False.

```
        if self.wait:
            wait_for(
                lambda: self.get_file_system_state(name),
                self.STATE_AVAILABLE,
                self.wait_timeout
            )
```

I'm assuming you've set wait to True in your test cases?
This is why it would always succeed on subsequent runs in my tests as it only took a few seconds to become available and that was at least the gap between runs.

I can think of a number of fixes, but not currently got time to try them out. 
Are there any scenarios where you wouldn't want to wait for the filesystem being available considering the short amount of time it takes?
Alternatively, before you create targets, you could add another wait to ensure the filesystem is available?

ryansydnor commented:
@jonhadfield - thanks for the debugging help! You were spot-on with `wait: True` in my test.

To answer your question: 

> Are there any scenarios where you wouldn't want to wait for the filesystem being available considering the short amount of time it takes?

No. 

I don't think the added complexity of a wait here is worthwhile. I've removed it and updated the PR. Let me know if things are working for you now. 

Thanks!

jonhadfield commented:
Creation now working, but deletion is only working by id and not name. 
Just returns without any errors.

ryansydnor commented:
@jonhadfield - I am not able to reproduce. Deletion with wait:yes and wait:no using both ID and name are working for me. Can you show me what you're doing that is failing?

Code snippets:

```
- name: Setting test efs instance name
  set_fact:
    efs_instance_name: "efs-test-{{ 99999 | random | to_uuid }}"

- name: Creating test EFS instance
  efs:
    state: present
    name: "{{ efs_instance_name }}"
    wait: yes
  register: result

- name: Saving ID of EFS instance
  set_fact:
    efs_instance_id: "{{ result.efs.FileSystemId }}"

- name: Deleting test EFS instance
  efs:
    state: absent
    id: "{{ efs_instance_id }}"
    #name: "{{ efs_instance_name }}"
    wait: no
    #wait: yes
  register: result
```

ryansydnor commented:
@gregdek - is there anything else I can do to push this along outside of wait for community members to test and sign off?

ryansb commented:
Works for me - :ship: ship_it :ship:

ryansb commented on cloud/amazon/efs.py:
```diff
@@ -0,0 +1,691 @@
+        description:
+            - Allows to create, search and destroy Amazon EFS file system
+        required: true
+        choices: ['present', 'absent', 'list']
```
I think `list` would be better off as its own `facts` module, not a "state" on this one.

ryansydnor replied:
done.

ryansb commented on cloud/amazon/efs.py:
```diff
@@ -0,0 +1,691 @@
+
+'''
+
+RETURN = '''
```
For returned values, please use [camel_dict_to_snake_dict](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ec2.py#L282) to match the variable style returned by other modules.

ryansydnor replied:
what about for the "targets" interface which is expecting a list of dicts with { SubnetId, IpAddress, SecurityGroups }?

is this a hard requirement? it'll require a somewhat significant refactor at this point.

ryansb replied:
The camel2snake function is for return values primarily, and recursive, so you just have to pass your data through it like `exit_json(changed=True, **camel_dict_to_snake_dict(your_stuff))`. 

For the input dictionary `targets` it should be pretty straightforward to switch the cases, something like this dict comprehension would do:

```
translations = {'ip_address': 'IpAddress',
 'security_groups': 'SecurityGroups',
 'subnet_id': 'SubnetId'}
new_targets = {translations[k]: v for k,v in targets.items()}
```

That'd take a snake-case version of the targets parameter and translate it to the camel-case version the AWS API wants, so you would just have to add that bit when taking options in.

ryansb commented on cloud/amazon/efs.py:
```diff
@@ -0,0 +1,691 @@
+            file_systems_info = filter(lambda item: has_tags(item['Tags'], tags), file_systems_info)
+
+        if targets:
+            targets = map(lambda item: (item, prefix_to_attr(item)), targets)
```
Minor style nit: it's preferred to use listcomps rather than map/filter, so this would become: `[prefix_to_attr(item) for item in targets]`

ryansydnor replied:
done.

gregdek commented:
@ryansydnor A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

ryansydnor commented:
@gregdek - addressed the comments from ryansb. split the "list" functionality into its own module (included in this PR for testing purposes).

tima commented:
ready_for_review

ryansb commented on cloud/amazon/efs.py:
```diff
@@ -0,0 +1,677 @@
+            - Allows to create, search and destroy Amazon EFS file system
+        required: true
+        choices: ['present', 'absent']
+        aliases: []
```
For future reference, anytime `aliases` is empty you can omit it.

ryansb commented on cloud/amazon/efs_facts.py:
```diff
@@ -0,0 +1,423 @@
+'''
+
+RETURN = '''
+CreationTime:
```
Same comment here about camel2snake.

ryansb commented:
Still needs to do proper argument handling with camel/snake case. Because that's the API we expose to users, we can't easily expose it as camelcase then switch it later.

ryansb commented:
Still needs to do proper argument handling with camel/snake case. Because that's the API we expose to users, we can't easily expose it as camelcase then switch it later.

gregdek commented:
@ryansydnor Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

ryansydnor commented:
@ryansb - updated with snake case. Let me know if it looks good!

tima commented:
ready_for_review

ryansb commented on cloud/amazon/efs.py:
```diff
@@ -0,0 +1,615 @@
+    """
+     Method creates iterator from boto result set
+    """
+    args = dict((key, value) for (key, value) in kwargs.iteritems() if value is not None)
```
You need to use `items` here, not `iteritems` to properly support Python 2.6.

ryansydnor commented:
ready_for_review

gregdek commented:
Thanks @ryansydnor for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

fvant commented:
Can state please be optional and default to present as with other modules ?

And can you add the performanceMode parameter too ?

fvant commented on cloud/amazon/efs_facts.py:
```diff
@@ -0,0 +1,378 @@
+
+DOCUMENTATION = '''
+---
+module: efs
```
efs_facts

ryansydnor replied:
Done. Thanks for pointing that out.

fvant commented on cloud/amazon/efs_facts.py:
```diff
@@ -0,0 +1,378 @@
+    type: str
+    sample: creating, available, deleting, deleted
+mount_point:
+    description: url of file system
```
not the complete url,  you need to prefix it with the appropriate AZ name.

ryansydnor replied:
Added appropriate comment mentioning the lack of AZ in the URL

ryansydnor commented:
@fvant - state is now optional (default to present). added performance mode as a parameter.

thanks again for taking a look!

fvant commented:
shipit

paprins commented:
LGTM

tima commented:
@paprins: The ansibots doesn't understand LGTM. ;)

shipit

gregdek commented:
Thanks again to @ryansydnor for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

ryansb commented on cloud/amazon/efs_facts.py:
```diff
@@ -0,0 +1,383 @@
+
+    def __init__(self, module, region, **aws_connect_params):
+        try:
+            session = Session(
```
You need to be using boto3_conn from EC2 module_utils, this connection method doesn't work for users that have multiple AWS profiles set up.

ryansb commented on cloud/amazon/efs_facts.py:
```diff
@@ -0,0 +1,383 @@
+    file_systems_info = connection.get_file_systems(FileSystemId=fs_id, CreationToken=name)
+
+    if tags:
+        file_systems_info = filter(lambda item: has_tags(item['Tags'], tags), file_systems_info)
```
i'd really prefer `[item for item in file_systems_info if has_tags(item['Tags'], tags)]` here. Filter with lambdas is a bit clunky.

ryansb commented on cloud/amazon/efs_facts.py:
```diff
@@ -0,0 +1,383 @@
+    if targets:
+        targets = [(item, prefix_to_attr(item)) for item in targets]
+        file_systems_info = filter(lambda item:
+                                   has_targets(item['MountTargets'], targets), file_systems_info)
```
Same here - list comprehension would be cleaner.

ryansb commented on cloud/amazon/efs_facts.py:
```diff
@@ -0,0 +1,383 @@
+                                   has_targets(item['MountTargets'], targets), file_systems_info)
+
+    file_systems_info = [camel_dict_to_snake_dict(x) for x in file_systems_info]
+    module.exit_json(changed=False, efs=file_systems_info)
```
For facts modules we encourage returning your info as part of `ansible_facts` so users don't have to register your output, for example:

```
module.exit_json(changed=False, ansible_facts={'efs': file_systems_info})
```

That way usage could look like:

```
- efs_facts:
    profile: my-aws-account
- command: "echo my EFS has {{ efs.number_of_mount_targets }} targets associated"
```

ryansb commented on cloud/amazon/efs.py:
```diff
@@ -0,0 +1,633 @@
+            CreationToken=name,
+            FileSystemId=file_system_id
+        ))
+        return info and info['LifeCycleState'] or self.STATE_DELETED
```
To avoid this boolean you can instead use `info.get('LifeCycleState', self.STATE_DELETED)`

ryansb commented on cloud/amazon/efs.py:
```diff
@@ -0,0 +1,633 @@
+        'security_groups': 'SecurityGroups',
+        'subnet_id': 'SubnetId'
+    }
+    targets = [dict((target_translations[key], value) for (key, value) in x.items()) for x in module.params.get('targets')]
```
Can you add comments to explain what's going on here and what data goes in/out of this?

ryansb commented on cloud/amazon/efs.py:
```diff
@@ -0,0 +1,633 @@
+
+    def __init__(self, module, region, **aws_connect_params):
+        try:
+            session = Session(
```
Same boto3_conn connection comment as below. Just replace with:

``` patch
         try:
-            session = Session(
-                aws_access_key_id=aws_connect_params['aws_access_key_id'],
-                aws_secret_access_key=aws_connect_params['aws_secret_access_key'],
-                aws_session_token=aws_connect_params['aws_session_token'],
-                region_name=region
-            )
-            self.connection = session.client('efs')
+            self.connection = boto3_conn(module, conn_type='client',
+                                         resource='efs', region=region,
+                                         **aws_connect_params)
         except Exception as e:
-            module.fail_json(msg=repr(e))
+            module.fail_json(msg="Failed to connect to AWS " + str(e), exception=traceback.format_exc(e))
```

ryansydnor commented:
@gregdek & @ryansb - what's going on with the waiting_on_contributor label? 

Are ryansb's most recent comments blocking? Or should we move forward and revisit later?

ryansb commented:
Going to go ahead with shipping this, but I'd really appreciate if you could follow up on these changes in another PR

ryansb commented:
Merged as https://github.com/ansible/ansible-modules-extras/commit/412af426c4a7ed4d3f3d271b87e4f5a09154a62c and https://github.com/ansible/ansible-modules-extras/commit/c03e26bd20efc494868be4cd3ae90e7102a60f5d

tima commented:
Nice work @ryansydnor. Thanks.

wimnat commented:
@ryansydnor @ryansb state should also be explicit not optional - core team request from a while back.  Also, there isn't a single try catch block around any of the boto3 calls.... :(

tima commented:
@wimnat I see https://github.com/ansible/ansible-modules-extras/pull/2305#issuecomment-246670832. Where's the core team request from awhile ago?

wimnat commented:
@tima i had a look around and couldn't find it.  It'll be either in the google devel group or a comment on a PR somewhere!  Pretty much untraceable but i didn't imagine it :)

I'll do a PR to update the docs

gundalow commented:
Did a follow up PR get raised, I don't see a link here
